opentelemetry-sdk/src/log/record.rs

```diff
@@ -0,0 +1,298 @@
+use opentelemetry_api::trace::{SpanContext, SpanId, TraceContextExt, TraceFlags, TraceId};
+use std::{borrow::Cow, collections::BTreeMap, time::SystemTime};
+
+#[derive(Debug, Clone, Default)]
+#[non_exhaustive]
+/// LogRecord represents all data carried by a log record, and
+/// is provided to `LogExporter`s as input.
+pub struct LogRecord {
+    /// Record timestamp
+    pub timestamp: Option<SystemTime>,
+
+    /// Timestamp for when the record was observed by OpenTelemetry
+    pub observed_timestamp: Option<SystemTime>,
+
+    /// Trace context for logs associated with spans
+    pub trace_context: Option<TraceContext>,
+
+    /// The original severity string from the source
+    pub severity_text: Option<Cow<'static, str>>,
+    /// The corresponding severity value, normalized
+    pub severity_number: Option<Severity>,
+
+    /// Record body
+    pub body: Option<Any>,
+
+    /// Resource attributes for the entity that produced this record
+    pub resource: Option<BTreeMap<Cow<'static, str>, Any>>,
+    /// Additional attributes associated with this record
+    pub attributes: Option<BTreeMap<Cow<'static, str>, Any>>,
+}
+
+impl LogRecord {
+    /// Create a [`LogRecordBuilder`] to create a new Log Record
+    pub fn builder() -> LogRecordBuilder {
+        LogRecordBuilder::new()
+    }
+}
+
+/// TraceContext stores the trace data for logs that have an associated
+/// span.
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct TraceContext {
```

Comment:
I think it refers to [W3C trace context standard](https://github.com/open-telemetry/opentelemetry-specification/blob/76e618535d841451b310f51522996362a6c1ec8f/specification/logs/data-model.md#trace-context-fields). Maybe we can merge it with the definition in `trace` mod?

Reply:
Yeah, just confusing as the concept is used multiple places, like the [TraceContextPropagator](https://docs.rs/opentelemetry/latest/opentelemetry/sdk/propagation/struct.TraceContextPropagator.html#) doesn't use this and the [TraceContextExt](https://docs.rs/opentelemetry/latest/opentelemetry/trace/trait.TraceContextExt.html) doesn't involve it either (maybe the ext should be renamed as well to disambiguate?)

Reply:
Maybe easier to divide the trace context field into trace id, span id, and trace flag and use the basic types. Or alternatively, those fields should only be present when `trace` is enabled. so that we can just refer the types from `trace` mod

Reply:
There is [`SpanContext`](https://docs.rs/opentelemetry/latest/opentelemetry/trace/struct.SpanContext.html#method.span_id), which stores the span ID, trace ID, and trace flags. However, it also stores an additional `TraceState` field, which is not in a Log Record. That's why I had to create this field.